zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/remote/ExecRemoteInterpreterProcess.java

```diff
@@ -0,0 +1,234 @@
+      cmdLine.addArgument(userName, false);
+    }
+    cmdLine.addArgument("-l", false);
+    cmdLine.addArgument(getLocalRepoDir(), false);
+    cmdLine.addArgument("-g", false);
+    cmdLine.addArgument(getInterpreterSettingName(), false);
+
+    interpreterProcessLauncher = new InterpreterProcessLauncher(cmdLine, getEnv());
+    interpreterProcessLauncher.launch();
+    interpreterProcessLauncher.waitForReady(getConnectTimeout());
+    if (interpreterProcessLauncher.isLaunchTimeout()) {
+      throw new IOException(
+          String.format("Interpreter Process creation is time out in %d seconds", getConnectTimeout() / 1000) + "\n"
+              + "You can increase timeout threshold via "
+              + "setting zeppelin.interpreter.connect.timeout of this interpreter.\n"
+              + interpreterProcessLauncher.getErrorMessage());
+    }
+
+    if (!interpreterProcessLauncher.isRunning()) {
+      throw new IOException("Fail to launch interpreter process:\n" + interpreterProcessLauncher.getErrorMessage());
+    } else {
+      String launchOutput = interpreterProcessLauncher.getProcessLaunchOutput();
+      Matcher m = YARN_APP_PATTER.matcher(launchOutput);
+      if (m.find()) {
+        String appId = m.group(1);
+        LOGGER.info("Detected yarn app: {}, add it to YarnAppMonitor", appId);
+        YarnAppMonitor.get().addYarnApp(ConverterUtils.toApplicationId(appId), this);
+      }
+    }
+  }
+
+  @Override
+  public void processStarted(int port, String host) {
+    super.processStarted(port, host);
+    // for yarn cluster it may be transitioned from COMPLETED to RUNNING.
+    interpreterProcessLauncher.onProcessRunning();
+  }
+
+  @Override
+  public void stop() {
+    if (isRunning()) {
+      super.stop();
+      // wait for a clean shutdown
+      this.interpreterProcessLauncher.waitForShutdown(RemoteInterpreterServer.DEFAULT_SHUTDOWN_TIMEOUT + 500);
+      // kill process
+      this.interpreterProcessLauncher.stop();
+      this.interpreterProcessLauncher = null;
+      LOGGER.info("Remote exec process of interpreter group: {} is terminated", getInterpreterGroupId());
+    } else {
+      LOGGER.warn("Try to stop a not running interpreter process of interpreter group: {}", getInterpreterGroupId());
+    }
+  }
+
+  @VisibleForTesting
+  public String getInterpreterRunner() {
+    return interpreterRunner;
+  }
+
+  @Override
+  public boolean isRunning() {
+    return interpreterProcessLauncher != null && interpreterProcessLauncher.isRunning();
+  }
+
+  @Override
+  public String getErrorMessage() {
+    return this.interpreterProcessLauncher != null
+        ? this.interpreterProcessLauncher.getErrorMessage()
+        : "";
+  }
+
+  private class InterpreterProcessLauncher extends ProcessLauncher {
+
+    public InterpreterProcessLauncher(CommandLine commandLine, Map<String, String> envs) {
+      super(commandLine, envs);
+    }
+
+    public void waitForShutdown(int timeout) {
```

Comment:
What is this waitForShudown's purpose ?  Wait for the process running and then shutdown ?

Reply:
The purpose is that Zeppelin server waits for a clean shutdown of the remote interpreter process.
Application flow:
 - Zeppelin interpreter process must be stopped, reason can be different
   - [code line](https://github.com/apache/zeppelin/pull/4035/files#diff-0f9e2999402c639082389ae5e0f54dbf631ce099f877850d8a7db47346a00000R79) 
 - Zeppelin Server performs a shutdown call, which is a synchronous operation
   -  [```
         callRemoteFunction(client -> {
        client.shutdown();
        return null;
      });```](https://github.com/apache/zeppelin/pull/4035/files#diff-0f9e2999402c639082389ae5e0f54dbf631ce099f877850d8a7db47346a00000R82)
 - The shutdown of the remote interpreter itself is asynchronous, because it's done in a separate shutdown thread
   - [code line](https://github.com/apache/zeppelin/blob/a22b59e18a3ea9e844f7fb288c1c3bb9b0dc94b9/zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterServer.java#L259-L260)
 - Normally, the remote interpreter stops cleanly in only a few milliseconds
 - That the process is stopped is recognised by the `ExecuteResultHandler`, which is implemented by the `ProcessLauncher` and overridden by the `InterpreterProcessLauncher`

If the remote interpreter does not stop in a certain time, the Zeppelin server terminates the process as before this PR with the help of the [Watchdog](https://github.com/apache/zeppelin/blob/a22b59e18a3ea9e844f7fb288c1c3bb9b0dc94b9/zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/util/ProcessLauncher.java#L160-L165).

Reply:
Thanks for the explanation, this change make sense to me.